File: backend/app/schemas/invoice.py
import uuid
from datetime import datetime, date
from decimal import Decimal

from pydantic import BaseModel


class InvoiceLineItemResponse(BaseModel):
    id: uuid.UUID
    invoice_id: uuid.UUID
    line_number: int
    description: str
    hsn_sac_code: str | None
    quantity: Decimal | None
    unit: str | None
    unit_price: Decimal | None
    amount: Decimal
    gst_rate: Decimal | None
    cgst_amount: Decimal | None
    sgst_amount: Decimal | None
    igst_amount: Decimal | None
    created_at: datetime

    model_config = {"from_attributes": True}


class InvoiceResponse(BaseModel):
    id: uuid.UUID
    client_id: uuid.UUID
    invoice_type: str
    file_name: str
    file_size_bytes: int
    invoice_number: str | None
    invoice_date: date | None
    seller_name: str | None
    seller_gstin: str | None
    buyer_name: str | None
    buyer_gstin: str | None
    place_of_supply: str | None
    taxable_amount: Decimal | None
    cgst_amount: Decimal | None
    sgst_amount: Decimal | None
    igst_amount: Decimal | None
    total_tax: Decimal | None
    total_amount: Decimal | None
    processing_status: str
    processing_error: str | None
    ocr_confidence: Decimal | None
    recon_status: str
    tally_sync_status: str
    is_active: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class InvoiceDetailResponse(InvoiceResponse):
    line_items: list[InvoiceLineItemResponse] = []


class InvoiceUpdate(BaseModel):
    invoice_number: str | None = None
    invoice_date: date | None = None
    invoice_type: str | None = None
    seller_name: str | None = None
    seller_gstin: str | None = None
    buyer_name: str | None = None
    buyer_gstin: str | None = None
    place_of_supply: str | None = None
    taxable_amount: Decimal | None = None
    cgst_amount: Decimal | None = None
    sgst_amount: Decimal | None = None
    igst_amount: Decimal | None = None
    total_tax: Decimal | None = None
    total_amount: Decimal | None = None

File: frontend/lib/types/invoice.ts
export interface InvoiceLineItem {
  id: string;
  invoice_id: string;
  line_number: number;
  description: string;
  hsn_sac_code: string | null;
  quantity: number | null;
  unit: string | null;
  unit_price: number | null;
  amount: number;
  gst_rate: number | null;
  cgst_amount: number | null;
  sgst_amount: number | null;
  igst_amount: number | null;
  created_at: string;
}

export interface Invoice {
  id: string;
  client_id: string;
  invoice_type: "sales" | "purchase";
  file_name: string;
  file_size_bytes: number;
  invoice_number: string | null;
  invoice_date: string | null;
  seller_name: string | null;
  seller_gstin: string | null;
  buyer_name: string | null;
  buyer_gstin: string | null;
  place_of_supply: string | null;
  taxable_amount: number | null;
  cgst_amount: number | null;
  sgst_amount: number | null;
  igst_amount: number | null;
  total_tax: number | null;
  total_amount: number | null;
  processing_status: "pending" | "processing" | "completed" | "failed";
  processing_error: string | null;
  ocr_confidence: number | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface InvoiceDetail extends Invoice {
  line_items: InvoiceLineItem[];
}
